### PR TITLE
move: improve tests for safe

### PR DESCRIPTION
**Pull request by rvantonder**
Stacked on https://github.com/MystenLabs/sui/pull/8397.

Exercises previously uncovered functions in `safe.move`:

- `safe::withdraw` -- for this we update an existing test for `withdraw_` (which is wrapped in `withdraw`) plus an assertion on the balance

- `safe::self_revoke_transfer_capability`

### CODE REVIEW

**Comment by amnn** [comment]
Heads up, usually the pattern when we stack commits is to base everything off `main` so that later PRs in the stack.  It means that later PRs include commits from earlier PRs in the stack, which is annoying/confusing, but it means that you're able to land prefixes of the stack, while later PRs are still getting reviewed, which can speed things up.

The PR itself is much easier to review the way you have things set-up (with the base of this PR being the previous PR in the stack), but what happens when you want to land this one -- will you need to land it against the prefix, and then land all of that on `main`, or is it possible to change the `base` to `main` after the review?

File: crates/sui-framework/tests/safe_tests.move
Comment (amnn) [inline crates/sui-framework/tests/safe_tests.move:78]: nit: test against `initial_funds`?

```suggestion
        assert!(coin::value(&withdrawn_coin) == initial_funds, 0);
```

File: crates/sui-framework/tests/safe_tests.move
Comment (amnn) [inline crates/sui-framework/tests/safe_tests.move:167]: Not really related to your test, but why does `self_revoke_transfer_capability` accept the capability by reference, rather than by value, to make it impossible at a type system level to try and withdraw using a revoked capability?

**Review by amnn — APPROVED** [review]

**Comment by rvantonder** [comment]
> The PR itself is much easier to review the way you have things set-up (with the base of this PR being the previous PR in the stack)

Indeed, this is how I self-manage stacked PRs on GH and the benefit is primarily for review and secondarily for chopping up changes when it makes sense, and my own sanity.

> but what happens when you want to land this one -- will you need to land it against the prefix, and then land all of that on main, or is it possible to change the base to main after the review?

The workflow is:

1. I merge bottom of stack
2. Next PR base changes to be on top of `main` by itself (more details below). I rebase this PR on top of `main` (force push)
3. Hit automerge / wait for CI / merge
4. Go to (2) if I have more PRs in the stack

In this workflow 1 PR always equals 1 (squashed) commit, every PR is merged onto `main`, no PRs are merged into the base or non-main branch. The branch basing is strictly for ease of review.

So there's a manual step here where I rebase + force push (I have a local script to help with this). So it's a little bit of effort that I take on my part, that improves the overall objective function (easy to review and split code). 

I tend to wait for all things in the stack to be approved/green before starting to merge the stack. If I start merging early, or significant changes are needed in the middle of a PR, things up the stack need to be fixed up (the whole rebase workflow above). I try not to stack beyond 3 or so PRs (and usually don't need to) unless they are very simple, so it's manageable. 

> is it possible to change the base to main after the review?

Yes. In fact, GH will _automatically_ change a subsequent PR base branch to `main` if the brnach below it in the stack is merged into `main`. This happened [circa 2020](https://github.blog/changelog/2020-05-19-pull-request-retargeting/). But, GH just changing the base branch doesn't guarantee it's safe to merge. For dependent changes you need to rebase/force-push (the manual step above) after merging. 

I think GH has some UI buttons to do branch rebases now but I prefer pure `git` primitives and CLIs to do this. I've also tried graphite but I'm a bit stuck in my old ways 😄 GH merge queues purportedly also help manage this automatically, but I haven't tried it through that workflow yet.

**Comment by rvantonder** [comment]
> usually the pattern when we stack commits is to base everything off main so that later PRs in the stack. It means that later PRs include commits from earlier PRs in the stack, which is annoying/confusing,

Got it, I'm aware of this but find this very confusing as a reviewer, especially if the changes depend on each other, I don't want to subject folks to it.

> but it means that you're able to land prefixes of the stack, while later PRs are still getting reviewed, which can speed things up.

It's possible to still land prefixes of the stack with the workflow I described above without slowing down review. Later PRs can get reviewed all the same (and less confusingly because they all reflect one commit that will be merged). The cost is the author needs to be OK with the rebase / merging management :-)

File: crates/sui-framework/tests/safe_tests.move
Comment (rvantonder) [reply crates/sui-framework/tests/safe_tests.move:78]: great catch!

File: crates/sui-framework/tests/safe_tests.move
Comment (rvantonder) [reply crates/sui-framework/tests/safe_tests.move:167]: I love that tests are raising questions about interface params though :P I'm not sure. From how this is used, could it be because `object::id(...)` expects a reference?

https://github.com/MystenLabs/sui/blob/2343a71b84edcf04e27384b7d5d9b06e9504e06e/crates/sui-framework/sources/safe.move#L160-L164

Comment (amnn) [reply crates/sui-framework/tests/safe_tests.move:167]: That's probably not the issue here, because if it were, you could take the capability by value, and take a reference of it locally.  One possibility is that it might be to do with the flow that allows a safe owner to revoke a transfer capability by ID.  If that happens, the capability will still exist, but is not usable.  Perhaps for uniformity, self-revoke is handled the same way?

It does seem strange to me that there is no way to burn the capability though.

**Comment by amnn** [comment]
> In fact, GH will *automatically* change a subsequent PR base branch to main if the branch below it in the stack is merged into main. This happened [circa 2020](https://github.blog/changelog/2020-05-19-pull-request-retargeting/). But, GH just changing the base branch doesn't guarantee it's safe to merge. For dependent changes you need to rebase/force-push (the manual step above) after merging.

🤯 TIL!

The "ever-growing stack" approach, also has the same manual overhead of rebases and force-pushes, by the way, so it sounds like what you're doing is a clear useability win -- I'm going to start adopting it!  It sounds like I'll need to work off feature-branches in MystenLabs/sui rather than my own fork, but that seems fine.

**Comment by amnn** [comment]
FWIW, I use [git-branchless](https://github.com/arxanas/git-branchless) to support some of the common stack-based operations from the CLI (mostly visualisaton and restacking).

**Pull request merged**
